### PR TITLE
Mass asset 

### PR DESCRIPTION
**Pull request by yarovyii**
Signed-off-by: Hlib Yarovyi <HlibYarovyi@eaton.com>

### CODE REVIEW

**Comment by yarovyii** [comment]
tested

File: src/create.cpp [hunk 1]
```diff
@@ -40,27 +41,52 @@ unsigned Create::run()
...
+        auto createdAsset = AssetManager::getItem(*ret);
+
+        if (!createdAsset) {
+            auditError(createdAsset.error());
+            continue;
+        }
+
+        createdName.append(createdAsset->name);
+        auditInfo("Request CREATE asset id {} SUCCESS"_tr, createdAsset->name);
+        someOk = true;
```
Comment (etnKlendathu) [inline src/create.cpp:85]: Is this was made by requiments? No errored assets return?

**Review by etnKlendathu — CHANGES_REQUESTED** [review]
Idea is good, but should be continued. My proposal is to add override function AssetManager::createAsset with serializationinfo (forward declared) and use this function in origin function.

**Comment by yarovyii** [comment]
build failed, because fty-asset changes not merged

File: src/create.cpp [hunk 2]
```diff
@@ -40,27 +41,52 @@ unsigned Create::run()
...
+    cxxtools::JsonDeserializer deserializer(jsonIn);
+    deserializer.deserialize(si);
 
-    if (!createdAsset) {
-        auditError(createdAsset.error());
-        throw rest::errors::Internal(createdAsset.error());        
-    } 
+    if (si.findMember("assets")) {
+        auto assetsJsonList = si.getMember("assets");
+        for (auto it = assetsJsonList.begin(); it != assetsJsonList.end(); ++it) {
```
[the remark this answers is not in the document]
Comment (etnKlendathu) [reply src/create.cpp:60]: I insist to use foreach loop

File: src/create.cpp [hunk 3]
```diff
@@ -40,27 +42,61 @@ unsigned Create::run()
...
-        throw rest::errors::Internal(ret.error());
+    cxxtools::SerializationInfo si;
+    std::stringstream           jsonIn;
+    try {
+        jsonIn << m_request.body();
+        cxxtools::JsonDeserializer deserializer(jsonIn);
+        deserializer.deserialize(si);
+    } catch (const std::exception& e) {
+        logError(e.what());
+        return HTTP_CONFLICT;
```
Comment (etnKlendathu) [inline src/create.cpp:53]: Why conflict? It should be internal error? No?

File: src/create.cpp [hunk 4]
```diff
@@ -40,27 +42,61 @@ unsigned Create::run()
...
     }
 
-    std::string assetJson = m_request.body();
-
-    auto ret = AssetManager::createAsset(assetJson, user.login());
-    if (!ret) {
-        auditError(ret.error());
-        throw rest::errors::Internal(ret.error());
+    cxxtools::SerializationInfo si;
+    std::stringstream           jsonIn;
```
Comment (etnKlendathu) [inline src/create.cpp:46]: jsonIn(m_request.body());
And inside try block (otherwise this stream in parent scope...)

File: src/create.cpp [hunk 5]
```diff
@@ -40,27 +42,61 @@ unsigned Create::run()
...
+        auto assetsJsonList = si.getMember("assets");
+        for (auto it = assetsJsonList.begin(); it != assetsJsonList.end(); ++it) {
+
+            auto ret = AssetManager::createAsset(*it, user.login());
+            if (!ret) {
+                auditError(ret.error());
+                continue;
+            }
+
+            auto createdAsset = AssetManager::getItem(*ret);
```
Comment (etnKlendathu) [inline src/create.cpp:68]: Why are you check it this way? You are loading whole asset... but createAsset returns asset id and you could just simple get a name by id.

File: src/create.cpp [hunk 6]
```diff
@@ -40,27 +42,61 @@ unsigned Create::run()
...
+
+        createdName.append(createdAsset->name);
+        auditInfo("Request CREATE asset id {} SUCCESS"_tr, createdAsset->name);
+        someOk = true;
+    }
 
+    m_reply << *pack::json::serialize(createdName) << "\n\n";
 
-    return HTTP_OK;
+    return someOk ? HTTP_OK : HTTP_CONFLICT;
```
Comment (etnKlendathu) [inline src/create.cpp:99]: Again, why conflict? :)

**Review by etnKlendathu — CHANGES_REQUESTED** [review]

File: src/create.cpp (hunk 3 above)
Comment (yarovyii) [reply src/create.cpp:53]: Clement told me to do similar to delete.cpp. There you can find 'return HTTP_CONFLICT;'

Comment (etnKlendathu) [reply src/create.cpp:53]: But this is a strange...
@perrettecl please, resolve this :)

File: src/create.cpp [hunk 7]
```diff
@@ -40,25 +41,49 @@ unsigned Create::run()
...
-
-    if (!createdAsset) {
-        auditError(createdAsset.error());
-        throw rest::errors::Internal(createdAsset.error());        
-    } 
+    cxxtools::SerializationInfo assetsJsonList;
+    if (si.findMember("assets")) {
+        assetsJsonList = si.getMember("assets");
+    } else {
+        assetsJsonList.addMember(m_request.body());
```
Comment (etnKlendathu) [inline src/create.cpp:58]: strange, i didn't catch it, body is already parsed, but you adding to list.. mmm.... string? is it work?

File: src/create.cpp [hunk 8]
```diff
@@ -40,25 +41,49 @@ unsigned Create::run()
...
+        }
+
+        auto createdAsset = fty::asset::db::idToNameExtName(*ret);
+
+        if (!createdAsset) {
+            auditError(createdAsset.error());
+            continue;
+        }
+
+        createdName.append(createdAsset->first.c_str());
```
Comment (etnKlendathu) [inline src/create.cpp:77]: for c_str we should cut the hands :)

File: src/create.cpp [hunk 9]
```diff
@@ -40,25 +41,49 @@ unsigned Create::run()
...
+        auditInfo("Request CREATE asset id {} SUCCESS"_tr, createdAsset->first.c_str());
+        someOk = true;
+    }
 
-    auditInfo("Request CREATE asset id {} SUCCESS"_tr, createdAsset->name);
+    m_reply << *pack::json::serialize(createdName) << "\n\n";
 
+    if (!someOk) {
+        throw rest::errors::Internal("Some of assets creation FAILED");
+    }
```
Comment (etnKlendathu) [inline src/create.cpp:86]: if not someOk, then list is empty, right? why are you serialize it to output? Exists any reason?

File: src/create.cpp [hunk 10]
```diff
@@ -40,25 +34,57 @@ unsigned Create::run()
...
-
-    if (!createdAsset) {
-        auditError(createdAsset.error());
-        throw rest::errors::Internal(createdAsset.error());        
-    } 
-
-    m_reply << "{ " << utils::json::jsonify("id", createdAsset->name) << " }\n\n";
+    bool             someOk = false;
+    pack::StringList createdName;
+    auto             validate = [&](const uint32_t& id) {
```
Comment (etnKlendathu) [inline src/create.cpp:49]: To be honest I don't understand this function, name is - validate, but it insert result into return list.

File: src/create.cpp [hunk 11]
```diff
@@ -40,25 +34,57 @@ unsigned Create::run()
...
+        auto createdAsset = fty::asset::db::idToNameExtName(id);
+
+        if (!createdAsset) {
+            auditError(createdAsset.error());
+            return;
+        }
+
+        createdName.append(createdAsset->first);
+        auditInfo("Request CREATE asset id {} SUCCESS"_tr, createdAsset->first.c_str());
```
Comment (etnKlendathu) [inline src/create.cpp:58]: c_str!

File: src/create.cpp [hunk 12]
```diff
@@ -40,25 +34,57 @@ unsigned Create::run()
...
 
-    auto createdAsset = AssetManager::getItem(*ret);
-
-    if (!createdAsset) {
-        auditError(createdAsset.error());
-        throw rest::errors::Internal(createdAsset.error());        
-    } 
-
-    m_reply << "{ " << utils::json::jsonify("id", createdAsset->name) << " }\n\n";
+    bool             someOk = false;
```
Comment (etnKlendathu) [inline src/create.cpp:47]: Existence of someOk is under of question. You could just check if something is in list.

**Review by etnKlendathu — CHANGES_REQUESTED** [review]

**Review by etnKlendathu — APPROVED** [review]

**Review by perrettecl — APPROVED** [review]

**Pull request merged**